src/portal/src/app/project/immutable-tag/add-rule/add-rule.component.ts

```diff
@@ -0,0 +1,153 @@
+    styleUrls: ['./add-rule.component.scss']
+})
+export class AddRuleComponent implements OnInit, OnDestroy {
+    addRuleOpened: boolean = false;
+    @Output() clickAdd = new EventEmitter<ImmutableRetentionRule>();
+    @Input() rules: ImmutableRetentionRule[];
+    @Input() projectId: number;
+    metadata: RuleMetadate = new RuleMetadate();
+    rule: ImmutableRetentionRule = new ImmutableRetentionRule(this.projectId);
+    isAdd: boolean = true;
+    editRuleOrigin: ImmutableRetentionRule;
+    onGoing: boolean = false;
+    @ViewChild(InlineAlertComponent, { static: false }) inlineAlert: InlineAlertComponent;
+    constructor(private immutableTagService: ImmutableTagService) {
+
+    }
+
+    ngOnInit(): void {
+    }
+
+    ngOnDestroy(): void {
+    }
+
+    get repoSelect() {
+        return this.rule.scope_selectors.repository[0].decoration;
+    }
+
+    set repoSelect(repoSelect) {
+        this.rule.scope_selectors.repository[0].decoration = repoSelect;
+    }
+
+    set repositories(repositories) {
+        if (repositories.indexOf(",") !== -1) {
+            this.rule.scope_selectors.repository[0].pattern = "{" + repositories + "}";
+        } else {
+            this.rule.scope_selectors.repository[0].pattern = repositories;
+        }
+    }
+
+    get repositories() {
+        return this.rule.scope_selectors.repository[0].pattern.replace(/[{}]/g, "");
+    }
+
+    get tagsSelect() {
+        return this.rule.tag_selectors[0].decoration;
+    }
+
+    set tagsSelect(tagsSelect) {
+        this.rule.tag_selectors[0].decoration = tagsSelect;
+    }
+
+    set tagsInput(tagsInput) {
+        if (tagsInput.indexOf(",") !== -1) {
+            this.rule.tag_selectors[0].pattern = "{" + tagsInput + "}";
+        } else {
+            this.rule.tag_selectors[0].pattern = tagsInput;
+        }
+    }
+
+    get tagsInput() {
+        return this.rule.tag_selectors[0].pattern.replace(/[{}]/g, "");
+    }
+
+    canNotAdd(): boolean {
+        if (this.onGoing) {
+            return true;
+        }
+        if (!this.isAdd && compareValue(this.editRuleOrigin, this.rule)) {
+            return true;
+        }
+        return !(
+            this.rule.scope_selectors.repository[0].pattern
+            && this.rule.scope_selectors.repository[0].pattern.replace(/[{}]/g, "")
+            && this.rule.tag_selectors[0].pattern
+            && this.rule.tag_selectors[0].pattern.replace(/[{}]/g, ""));
+    }
+
+    open() {
+        this.addRuleOpened = true;
+        this.inlineAlert.alertClose = true;
+        this.onGoing = false;
+    }
+
+    close() {
+        this.addRuleOpened = false;
+    }
+
+    cancel() {
+        this.close();
+    }
+
+    add() {
+        // remove whitespaces
+        this.rule.scope_selectors.repository[0].pattern = this.rule.scope_selectors.repository[0].pattern.replace(/\s+/g, "");
+        this.rule.tag_selectors[0].pattern = this.rule.tag_selectors[0].pattern.replace(/\s+/g, "");
```

Comment:
show add validator to check if scope_selectors. 
For repository, it is the same.